File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let ht = {};
    
    let left = 0;
    let count = 0;
    
    let maxcount = 0;
    
    for(let i in s) {
        if(!ht[s[i]]) {
            ht[s[i]] = 1;
        } else {
            ht[s[i]]+=1;
        }
        
        count++;
        
        while(ht[s[i]] > 1) {
            ht[s[left]]--;
            left++;
            count--;
        }
        
        maxcount = Math.max(count, maxcount);
    }
    
    return maxcount;
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        myMap = {}
        
        count = 0
        
        maxCount = 0
        
        left = 0
        
        for r in range(len(s)):
            cur = s[r]

            if cur not in myMap:
                myMap[cur] = 0
            
            myMap[cur]+=1
            count+=1
            
            while myMap[cur] > 1: 
                myMap[s[left]] -= 1
                left+=1;
                count-=1
            
            maxCount = max(maxCount, count)
        
        return maxCount
                
            